### PR TITLE
Cloudfront deploy

### PR DESCRIPTION
**Pull request by sjezewski**
Update aws.sh script to enable cloudfront distribution.

Now if you run `./etc/testing/deploy/aws.sh --create --use-cloudfront`

it'll create a pachyderm cluster with:

- an s3 bucket w the correct bucket policy so that CF can access the bucket
- a cloudfront deployment
- with the correct caching rules so that it does NOT cache error responses (which pachd needs to see)
- pachd configured to use cloudfront

This PR does not:

- change cloudfront to be on by default
- update cloudfront to lock down access

### CODE REVIEW

**Comment by sjezewski** [comment]
Tested the fruit stand example ... seemed to work fine for the filter step ... the sum step reported an error ... but seems like its related to the transform ... seems like a bug w fruit stand?

```
  "transform": {
    "cmd": [ "sh" ],
    "stdin": [
        "for fruit in apple orange banana; do",
        "   [[ -e /pfs/filter/$fruit ]] || continue",
        "   cat /pfs/filter/$fruit | awk '{s+=$1} END {print s}' >/pfs/out/$fruit",
        "done"
    ]
  },
```

the error is:

```
17-06-04[18:16:11]:pachyderm:0$pc get-logs --job b9bd2040-bdda-40bc-9297-82ace8372775
sh: 2: [[: not found

sh: 2: [[: not found

sh: 2: [[: not found

sh: 2: [[: not found

sh: 2: [[: not found

sh: 2: [[: not found

sh: 2: [[: not found

sh: 2: [[: not found

sh: 2: [[: not found

```

File: etc/deploy/aws.sh [hunk 1]
```diff
@@ -1,18 +1,27 @@
...
+  which uuid
   # Common config
   export AWS_REGION=us-east-1
   export AWS_AVAILABILITY_ZONE=us-east-1a
   export STATE_BUCKET=s3://k8scom-state-store-pachyderm-${RANDOM}
   local USE_EXISTING_STATE_BUCKET='false'
 
   # Parse flags
-  eval "set -- $( getopt -l "state:,region:,zone:,no-metrics" "--" "${0}" "${@}" )"
+  eval "set -- $( getopt -l "state:,region:,zone:,no-metrics:,use-cloudfront" "--" "${0}" "${@}" )"
```
Comment (sjezewski) [inline etc/deploy/aws.sh]: Now that I've read up on getopts I realize `no-metrics` shouldn't have a trailing colon. Will fix.

File: etc/deploy/aws.sh [hunk 2]
```diff
@@ -29,7 +38,11 @@ parse_flags() {
             shift 2
             ;;
           --no-metrics)
-            METRICS_FLAG="false" # default is false, see top of file
+            METRICS_FLAG="false" # default is true, see top of file
```
Comment (sjezewski) [inline etc/deploy/aws.sh]: I believe this was a typo

**Comment by sjezewski** [comment]
Oh and I should mention that I tested `./etc/testing/deploy/aws.sh` with and without the `--use-cloudfront` flag and it worked fine.

Once cloudfront becomes the default (in the next PR) this deployment code path will be hooked into the nightly test suite.

File: etc/deploy/aws.sh (hunk 1 above)
Comment (msteffen) [inline etc/deploy/aws.sh]: :+1:

File: etc/deploy/aws.sh (hunk 2 above)
Comment (msteffen) [inline etc/deploy/aws.sh]: Yup, it was a typo

File: etc/deploy/aws.sh [hunk 3]
```diff
@@ -72,13 +85,33 @@ create_s3_bucket() {
   else
     aws s3api create-bucket --bucket ${BUCKET} --region ${AWS_REGION} --create-bucket-configuration LocationConstraint=${AWS_REGION}
   fi
+
+  if [ "$2" != "false" ]; then
```
Comment (msteffen) [inline etc/deploy/aws.sh]: Could you add a short comment to the top of `create_s3_bucket` indicating that it takes two arguments, and what the second argument means? (the first is the name of the s3 bucket being created)

File: etc/deploy/aws.sh [hunk 4]
```diff
@@ -220,14 +253,18 @@ deploy_pachyderm_on_aws() {
...
     # Since my user should have the right access:
     AWS_KEY=`cat ~/.aws/credentials | grep aws_secret_access_key | cut -d " " -f 3`
     AWS_ID=`cat ~/.aws/credentials | grep aws_access_key_id  | cut -d " " -f 3`
 
     # Omit token since im using my personal creds
-    cmd=( pachctl deploy amazon ${BUCKET_NAME} "${AWS_ID}" "${AWS_KEY}" " " ${AWS_REGION} ${STORAGE_SIZE} --dynamic-etcd-nodes=3 )
+    if [[ "$USE_CLOUDFRONT" == "true" ]]; then
+        cmd=( pachctl deploy amazon ${BUCKET_NAME} "${AWS_ID}" "${AWS_KEY}" " " ${AWS_REGION} ${STORAGE_SIZE} --dynamic-etcd-nodes=3 --cloudfront-distribution ${CLOUDFRONT_DOMAIN})
+    else
+        cmd=( pachctl deploy amazon ${BUCKET_NAME} "${AWS_ID}" "${AWS_KEY}" " " ${AWS_REGION} ${STORAGE_SIZE} --dynamic-etcd-nodes=3 )
```
Comment (msteffen) [inline etc/deploy/aws.sh]: This is entirely a matter of aesthetics, but how you feel about:
```
cmd=( pachctl deploy amazon ${BUCKET_NAME} "${AWS_ID}" "${AWS_KEY}" " " ${AWS_REGION} ${STORAGE_SIZE} --dynamic-etcd-nodes=3 --cloudfront-distribution ${CLOUDFRONT_DOMAIN})
if [[ "$USE_CLOUDFRONT" == "true" ]]; then
  cmd+=("--cloudfront-distribution ${CLOUDFRONT_DOMAIN}")
fi
if [[ "${METRICS_FLAG}" == "false" ]]; then  # Already there...
  cmd+=( "--no-metrics" )
fi
```

File: etc/testing/deploy/aws.sh [hunk 5]
```diff
@@ -57,7 +63,7 @@ set -x
 case "${OP}" in
   create)
     aws_sh="$(realpath "$(dirname "${0}")/../../deploy/aws.sh")"
-    sudo "${aws_sh}" --region=${REGION} --zone=${ZONE} --state=${STATE_STORE} --no-metrics
+    sudo "${aws_sh}" --region=${REGION} --zone=${ZONE} --state=${STATE_STORE} --no-metrics $CLOUDFRONT
```
Comment (msteffen) [inline etc/testing/deploy/aws.sh]: My memory is that passing an empty flag is not the same thing as passing no flag, so if this doesn't work, you may need to rely on a pattern like
```
cmd=("${aws_sh}" --region=${REGION} --zone=${ZONE} --state=${STATE_STORE} --no-metrics)
if [[ -n "${CLOUDFRONT}" ]]; then
  cmd+=("${CLOUDFRONT}")
fi
```

similar to above...

**Comment by msteffen** [comment]
Re: the fruit stand, that's a bug (probably my fault). `[[` is a bash-only enhancement of `[` which I'm too in the habit of using.

File: etc/testing/deploy/aws.sh (hunk 5 above)
Comment (sjezewski) [inline etc/testing/deploy/aws.sh]: It does work, but I'll make it consistent

File: etc/testing/deploy/aws.sh [hunk 6]
```diff
@@ -57,7 +63,11 @@ set -x
 case "${OP}" in
   create)
     aws_sh="$(realpath "$(dirname "${0}")/../../deploy/aws.sh")"
-    sudo "${aws_sh}" --region=${REGION} --zone=${ZONE} --state=${STATE_STORE} --no-metrics
+    cmd=("${aws_sh}" --region=${REGION} --zone=${ZONE} --state=${STATE_STORE} --no-metrics)
+    if [[ -n "${CLOUDFRONT}" ]]; then
+      cmd+=(" ${CLOUDFRONT}")
```
Comment (msteffen) [inline etc/testing/deploy/aws.sh]: I believe the leading space is unnecessary. The way bash arrays work is the parsing is done before the array is created, so by appending this to the array using this syntax you don't need a separator (as opposed to e.g. `cmd="${cmd[*]} ${CLOUDFRONT}"`)

File: etc/testing/deploy/aws.sh [hunk 7]
```diff
@@ -57,7 +63,11 @@ set -x
...
+    fi
+    sudo $cmd
```
Comment (msteffen) [inline etc/testing/deploy/aws.sh]: To expand the array here, I think you'll want to use:
`sudo "${cmd[@]}"`. That syntax expands `cmd` to all of its elements and quotes them, and I believe is generally the way people forward arguments. It's the same as "${@}" if you've ever used that idiom before)

**Comment by sjezewski** [comment]
The checks actually succeeded. I'm guessing this is a relic of the GH outage this afternoon.

Merging now.

**Pull request merged**